File: LeetCode/0049.group-anagrams/group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<HashableArray, List<String>>hashMap = new HashMap<>();
        for (String str : strs){
            HashableArray hashableArray = new HashableArray();
            for (char ch : str.toCharArray()){
                hashableArray.val[ch - 'a']++;
            }
            if (!hashMap.containsKey(hashableArray))    hashMap.put(hashableArray, new ArrayList<>());
            hashMap.get(hashableArray).add(str);
        }
        return new ArrayList<>(hashMap.values());
    }
}

class HashableArray{
    int[] val = new int[26];

    @Override
    public boolean equals(Object obj) {
        return Arrays.equals(this.val, ((HashableArray)obj).val);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(val);
    }
}

File: LeetCode/0049.group-anagrams/group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int size = strs.size();
        unordered_map<string,vector<string>>mp;
        vector<vector<string>> res;
        for (int i = 0; i < size; i++){
            string copy(strs[i]);
            sort(copy.begin(), copy.end());
            mp[copy].push_back(strs[i]);
        }
        for (unordered_map<string, vector<string>>::iterator it = mp.begin(); it != mp.end(); it++){
            res.push_back(it->second);
        }
        return res;
    }
};